File: python/binary-tree-general/pop_next_right_pointer.py
from collections import deque
from typing import Optional

class Node:
    def __init__(self, val: int = 0, left: Optional['Node'] = None, right: Optional['Node'] = None, next: Optional['Node'] = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: Optional[Node]) -> Optional[Node]:
        if not root:
            return None

        queue = deque([root])

        while queue:
            prev = None

            level_size = len(queue)

            for _ in range(level_size):
                node = queue.popleft()

                if prev:
                    prev.next = node

                prev = node


                if node.left:
                    queue.append(node.left)

                if node.right:
                    queue.append(node.right)


            prev.next = None

        return root

File: golang/binary-tree-general/pop_next_right_pointer.go
package main

import "container/list"

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	if root == nil {
		return nil
	}

	queue := list.New()

	queue.PushBack(root)

	for queue.Len() > 0 {
		prev := (*Node)(nil)
		levelSize := queue.Len()

		for i := 0; i < levelSize; i++ {
			element := queue.Front()

			node := element.Value.(*Node)

			queue.Remove(element)

			if prev != nil {
				prev.Next = node
			}

			prev = node

			if node.Left != nil {
				queue.PushBack(node.Left)
			}

			if node.Right != nil {
				queue.PushBack(node.Right)
			}
		}

		prev.Next = nil
	}

	return root
}
